theano/gpuarray/dnn.py

```diff
@@ -2800,6 +2800,204 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
     return [rval]
 
 
+class _GpuDnnTransformerDescriptor(COp):
+
+    """
+    This Op builds a spatial transformer descriptor for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dtype',)
+    params_type = ParamsType(dtype=cudnn.cudnnDataType_t)
+
+    def c_headers(self):
+        return ['cudnn.h', 'cudnn_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), config.dnn.include_path]
+
+    def c_libraries(self):
+        return ['cudnn']
+
+    def c_lib_dirs(self):
+        return [config.dnn.library_path]
+
+    def do_constant_folding(self, node):
+        return False
+
+    def __init__(self, dtype=theano.config.floatX):
+        COp.__init__(self, ["c_code/dnn_sptf_desc.c"], "APPLY_SPECIFIC(dnn_sptf_desc)")
+
+        assert cudnn.cudnnDataType_t.has_alias(dtype)
+        self.dtype = dtype
+
+    def make_node(self, dimensions):
+        # cuDNN supports only 2D transformations, and the output tensor must
+        # have exactly 4 dimensions: (num_images, num_channels, height, width)
+        assert len(dimensions) == 4
+        dimensions = tuple(dimensions)
+        nimages, nchannels, height, width = dimensions
+
+        node = Apply(self, [nimages, nchannels, height, width],
+                     [CDataType("cudnnSpatialTransformerDescriptor_t",
+                                freefunc="cudnnDestroySpatialTransformerDescriptor")()])
+        # DebugMode cannot compare the values of CDataType variables, so by
+        # default it returns False all the time. To prevent DebugMode from
+        # complaining because of the MergeOptimizer, we make this variable
+        # always compare to True.
+        out = node.outputs[0]
+        out.tag.values_eq_approx = tensor.type.values_eq_approx_always_true
+        return node
+
+    def c_code_cache_version(self):
+        return (super(_GpuDnnTransformerDescriptor, self).c_code_cache_version(), version())
+
+
+class GpuDnnTransformer(DnnBase):
+    """
+    This Op builds a spatial transformer that can be used in spatial transformer networks.
+    """
+
+    __props__ = ('dtype',)
+    _cop_num_inputs = 6
+    _cop_num_outputs = 2
+    _f16_ok = True
+    default_output = 0
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/dnn_sptf.c"], "dnn_sptf")
+        self.dtype = dtype
+
+    def make_node(self, img, theta, grid_dims, desc, alpha=None, beta=None):
```

Comment:
Most implementations in Theano use hardcoded values for ``alpha`` and ``beta`` (e.g. ``dnn_softmax``), do you think it is also the case here?

I don't think the user will want (or need) the grid of coordinates - generated by the forward grid, since it is only used by the sampler to map the pixels that must be sampled from the input images. Also, the only outputs the user is interested from the spatial transformer - in the forward pass, are the transformed images.

In the case of gradients, this is somewhat different, since the user will also want the gradients of the affine transformations - besides the gradients of the inputs, so it is possible to backpropagate them to the localization network. Therefore, there are separate classes for each of these gradients, which generated by separate backward functions of cuDNN. However, they are returned together in the ``L_op`` method of ``GpuDnnTransformer``, because the output of the backward function of the sampler is the input of the backward function of the grid generator.